clients/venice-client/src/main/java/com/linkedin/venice/fastclient/transport/GrpcTransportClient.java

```diff
@@ -22,209 +24,286 @@
+
     try {
       TlsChannelCredentials.Builder tlsBuilder = TlsChannelCredentials.newBuilder()
           .keyManager(GrpcUtils.getKeyManagers(sslFactory))
           .trustManager(GrpcUtils.getTrustManagers(sslFactory));
-      channelCredentials = tlsBuilder.build();
+      return tlsBuilder.build();
     } catch (Exception e) {
       throw new VeniceClientException(
           "Failed to initialize SSL channel credentials for Venice gRPC Transport Client",
           e);
     }
   }
 
-  protected ManagedChannel getChannel(String serverAddress) {
-    if (!nettyAddressToGrpcAddressMap.containsKey(serverAddress)) {
-      throw new VeniceException("No grpc server found for address: " + serverAddress);
+  @VisibleForTesting
+  VeniceClientRequest buildVeniceClientRequest(String[] requestParts, byte[] requestBody, boolean isSingleGet) {
+    VeniceClientRequest.Builder requestBuilder = VeniceClientRequest.newBuilder()
+        .setResourceName(requestParts[4])
+        .setIsBatchRequest(!isSingleGet)
+        .setMethod(isSingleGet ? HttpMethod.GET.name() : HttpMethod.POST.name());
+
+    if (isSingleGet) {
+      requestBuilder.setKeyString(requestParts[6]);
+      requestBuilder.setPartition(Integer.parseInt(requestParts[5]));
+    } else {
+      requestBuilder.setKeyBytes(ByteString.copyFrom(requestBody));
     }
 
-    return serverGrpcChannels.computeIfAbsent(
-        serverAddress,
-        k -> Grpc.newChannelBuilder(nettyAddressToGrpcAddressMap.get(k), channelCredentials).build());
+    return requestBuilder.build();
   }
 
-  protected VeniceReadServiceGrpc.VeniceReadServiceStub getStub(ManagedChannel channel) {
+  @VisibleForTesting
+  VeniceReadServiceGrpc.VeniceReadServiceStub getOrCreateStub(String serverAddress) {
+    String grpcAddress = getGrpcAddressFromServerAddress(serverAddress);
+
+    ManagedChannel channel = serverGrpcChannels
+        .computeIfAbsent(serverAddress, k -> Grpc.newChannelBuilder(grpcAddress, channelCredentials).build());
+
     return stubCache.computeIfAbsent(channel, VeniceReadServiceGrpc::newStub);
   }
 
-  @Override
-  public CompletableFuture<TransportClientResponse> get(String requestPath, Map<String, String> headers) {
-    return handleRequest(requestPath, headers, null, true);
+  /**
+   * Maps the given server address to GRPC server address. Typically, the mapped server address only differs in the
+   * port. Fetch the GRPC server address for a given server address if available in the #nettyServerToGrpcAddress
+   * map if not, strip the port from the original server address and replace it with the grpc port.
+   * <b>Note:</b> We have the GRPC port hard coded and operate with static port assignment for GRPC
+   */
+  @VisibleForTesting
+  String getGrpcAddressFromServerAddress(String serverAddress) {
+    Preconditions.checkState(StringUtils.isNotEmpty(serverAddress), "Server address cannot be empty ");
+
+    nettyServerToGrpcAddress.computeIfAbsent(serverAddress, nettyAddress -> {
```

Comment:
In most cases this method is just returning `serverAddress` + GRPC port. I guess one reason I can think of for still keeping the `nettyServerToGrpcAddress` map is to give us flexibility in case we have a few server hosts with different GRPC port.

Reply:
actually, the map only exists for the purpose of integration tests where ew bring up multiple venice servers and we need to map the netty server address to grpc server address (ports are dynamically generated).

Once, we add port discovery through metadata the map will go away